lib/analytics/index.js

```diff
@@ -0,0 +1,333 @@
+const GA_SERVER_PORT = '443';
+const GA_API_KEY = 'XuPojOTwQ6yTO758EV4hBg';
+const GA_TRACKING_ID = 'G-DEKPKZSLXS';
+
+const defaultSettings = {
+  enabled: false,
+  log_path: './logs/analytics',
+  client_id: uuid.v4()
+};
+
+const RESERVED_EVENT_NAMES = ['ad_activeview', 'ad_click', 'ad_exposure', 'ad_impression', 'ad_query', 'adunit_exposure',
+  'app_clear_data', 'app_install', 'app_update', 'app_remove', 'error', 'first_open', 'first_visit', 'in_app_purchase',
+  'notification_dismiss', 'notification_foreground', 'notification_open', 'notification_receive', 'os_update',
+  'screen_view', 'session_start', 'user_engagement'];
+
+
+const RESERVED_EVENT_PARAMETERS = ['firebase_conversion'];
+
+const MAX_PARAMETERS_LENGTH = 40;
+
+/**
+ * See: https://developers.google.com/analytics/devguides/collection/protocol/v1/devguide
+ */
+class AnalyticsCollector {  
+  constructor() {
+    this.queueLength = 0;
+    this.parameters = {};
+    this.logger = Logger;
+    this.initialized = false;
+
+    // This identifies unique users and helps us separate out sessions.
+    this.parameters['client_id'] = uuid.v4();
+
+    // Prevents google from indexing the events for ad targeting.
+    this.parameters['non_personalized_ads'] = true;
+  }
+
+  updateSettings(settings) {
+    this.settings = settings;
+
+    if (!this.settings.usage_analytics) {
+      this.settings.usage_analytics = defaultSettings;
+    }
+
+    // update client_id
+    this.parameters['client_id'] = this.settings.usage_analytics.client_id;
+
+    // update queueLength
+    if (this.settings.usage_analytics.enabled) {
+      try {
+        const stat = fs.statSync(this.__getLogFileLocation());
+        this.queueLength = stat.size / 400; // the actual value doesn't matter
+      } catch (e) {
+      // ignore
+      }
+    }
+
+    this.initialized = true;
+  }
+
+  updateLogger(logger) {
+    this.logger = logger;
+  }
+
+  event(name, parameters = {}) {
+    if (!this.initialized) {
+      throw Error('AnalyticsCollector is not initialized');
+    }
+
+    // Not all shape of events are supported.
+    this.__validateEvent(name, parameters);
+
+    // Add event timestamp.
+    parameters['event_time'] = new Date().getTime() * 1000;
+
+    // Add environment information.
+    parameters['env_os'] = buildUserAgentString();
+    parameters['env_language'] = getLanguage();
+    parameters['env_nw_version'] = VERSION.full;
+    parameters['env_node_version'] = `node ${process.version}`;
+
+    // Add call location.
+    parameters.callLocation = getCallLocation();
+
+    return this.__addToQueue(name, parameters);
+  }
+ 
+  exception(error, location = 'lib/runner/process-listener.js#uncaught') {
+    try {
+      const parameters = {
+        env_os: buildUserAgentString(),
+        env_language: getLanguage(),
+        env_nw_version: VERSION.full,
+        env_node_version: `node ${process.version}`,
+        modulePath: location,
+        errorMessage: error.message, 
```

Comment:
Can we change this to 
```suggestion
        errorMessage: error.name, 
```